### PR TITLE
feat: Allow per user logout url

### LINKED ISSUE #20285: Allow per-user logoutURL
**Is your feature request related to a problem? Please describe.**
Front-end applications may have users enter through different paths in the application, and when the user log's out want to return them to their previous login screen.

Currently, the `logoutURL` is shared by all users and doesn't provide any way to redirect different users to different areas.

**Describe the solution you'd like**
In the [join](https://docs.bigbluebutton.org/3.0/development/api/#join) API parameters, add support for a `logoutURL` to enable the front-end to send each user their own logout URL.

**Describe alternatives you've considered**
We could add parameters to the `logoutURL` on create, such as `%%USERID%%` or `%%ROLE%%`, but giving each user their own `logoutURL` is simpler and gives all the control to the front-end.

### PR DESCRIPTION
**Pull request by jaapjansma**
### What does this PR do?

This PR adds a logoutURL parameter to the joinMeeting api. This introduces a feature to have a per user logoutURL

### Closes Issue(s)

Closes #20285 

### Motivation

We are using big blue button in our organization and we create a meeting from our CiviCRM installation. We want the moderator to go to a different url after logging out.  In the url for the moderator we ask the moderator some follow up questions about the meeting and we store that result in our CiviCRM.

The viewers should go to the normal logout page. 

### More

- [ v ] Added/updated documentation:  yes, see #21205

### CODE REVIEW

**Comment by gustavotrott** [comment]
Hey @jaapjansma, thanks for your contribution!

Right now, the `logoutURL` supports placeholders like `%%USERID%%`, but I know these might not be super helpful for third-party integrations. 
I was thinking we could add new placeholders like `%%EXTERNAL_USERID%%` and maybe `%%USER_ROLE%%`. This way, the integration platform could use a callback endpoint and set it as the `logoutUrl`, which would handle the user properly and redirect them to the right page.

Basically, instead of passing the URL directly through the join call, your CRM would have a generic URL that gets called, receives a parameter with the user's ID in that application, and it knows what to do next.

Do you think this approach would work for your use case?

**Comment by jaapjansma** [comment]
@gustavotrott the short answer is that we might get that working with a little workaround. The reason for that is that our CiviCRM runs in a private network (behind a firewall) and the follow up form is on a separate website. Your suggestion means that we have to implement something in between which looks up the use role and then redirect it to the desired location (which depends on whether the person has the role moderator). 

So ideally we provide this logout Url from our CiviCRM as soon as we create the meeting. 

I am also curious what @ffdixon would think? As he has written the original feature request.

File: bbb-common-web/src/main/java/org/bigbluebutton/api2/IBbbWebApiGWApp.java [hunk 1]
```diff
@@ -74,7 +74,7 @@ void createMeeting(String meetingID, String externalMeetingID,
   void registerUser(String meetingID, String internalUserId, String fullname, String role,
                     String externUserID, String authToken, String sessionToken, String avatarURL, String webcamBackgroundURL,
                     Boolean guest, Boolean authed, String guestStatus, Boolean excludeFromDashboard,
-                    String enforceLayout, Map<String, String> userMetadata);
+                    String enforceLayout, String logoutUrl, Map<String, String> userMetadata);
```
[the remark this answers is not in the document]
Comment (jaapjansma) [reply bbb-common-web/src/main/java/org/bigbluebutton/api2/IBbbWebApiGWApp.java:77]: I have checked all implementations of  `IBbbWebApiGWApp` and only found one (`BbbWebApiGWApp` and that one was already updated.

File: bbb-common-web/src/main/scala/org/bigbluebutton/api2/meeting/MeetingsManagerActor.scala [hunk 2]
```diff
@@ -18,7 +18,7 @@ case class AddUserSession(token: String, session: UserSession)
 case class RegisterUser(meetingId: String, intUserId: String, name: String, role: String,
                         extUserId: String, authToken: String, sessionToken: String, avatarURL: String, webcamBackgroundURL: String,
                         guest: Boolean, authed: Boolean, guestStatus: String, excludeFromDashboard: Boolean,
-                        enforceLayout: String, userMetadata: Map[String, String])
+                        enforceLayout: String, logoutUrl: String, userMetadata: Map[String, String])
```
[the remark this answers is not in the document]
Comment (jaapjansma) [reply bbb-common-web/src/main/scala/org/bigbluebutton/api2/meeting/MeetingsManagerActor.scala:21]: I understand this comment that code analyses thinks I replaced the `sessionToken` parameter. However I added logoutUrl paramater, just after `enforceLayout` and before `userMetaData`. So I do think the suggestions from this comment do not apply on this change.

**Comment by jaapjansma** [comment]
I have resolved the conflicting changes

**Comment by gustavotrott** [comment]
The build of `bbb-common-message` is not working!

File: bbb-common-message/src/main/scala/org/bigbluebutton/common2/util/UrlValidationUtil.scala [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+object UrlValidationUtil {
+  def isValidUrl(url: String): Boolean = {
+    if (url == null) return false
+    try {
+      val parsedUrl = new URL(url)
+      val protocol = parsedUrl.getProtocol
+      protocol == "http" || protocol == "https"
+    } catch {
+      case _: MalformedURLException => false
```
Comment (antobinary) [inline bbb-common-message/src/main/scala/org/bigbluebutton/common2/util/UrlValidationUtil.scala:13]: [error] /tmp/build/bbb-fsesl-akka_3.0.0~beta.4+0-git.0_jammy/bbb-common-message/src/main/scala/org/bigbluebutton/common2/util/UrlValidationUtil.scala:13:15: not found: type MalformedURLException

**Comment by antobinary** [comment]
Hi @jaapjansma, please mark the PR as ready to review when you are done with your changes. Thanks again for this work!!

**Comment by jaapjansma** [comment]
@antobinary I have done my work. I updated this PR to relfect the failing tests.

**Comment by gustavotrott** [comment]
I sent a PR with a fix to the failing automated-tests step, can you take a look @jaapjansma ?

https://github.com/jaapjansma/bigbluebutton/pull/1

**Comment by jaapjansma** [comment]
@gustavotrott thanks!

**Comment by gustavotrott** [comment]
The URL validation makes the placeholders like `%%USERNAME%%` stop working for `logoutURL` parameter.

Maybe we can accept the PR as is, and refactor the code to handle this cases. 
Once each user will have a prop `logoutUrl`, we can apply the placeholders values in the backend and the front don't need to handle it anymore.
Thoughts @lfzawacki ?

**Comment by jaapjansma** [comment]
If needed I am happy to create a follow up PR fixing this issue

**Comment by gustavotrott** [comment]
> If needed I am happy to create a follow up PR fixing this issue

That would be great!
Thank you

**Comment by lfzawacki** [comment]
> The URL validation makes the placeholders like `%%USERNAME%%` stop working for `logoutURL` parameter.
> 
> Maybe we can accept the PR as is, and refactor the code to handle this cases. Once each user will have a prop `logoutUrl`, we can apply the placeholders values in the backend and the front don't need to handle it anymore. Thoughts @lfzawacki ?

It makes sense, the reason I didn't do the substitution on the back end is because by the point we generated the logoutUrl we did not have access to all the data we needed. Doing it on a per user basis should make it possible.
